stake-pool/program/src/processor.rs

```diff
@@ -2548,11 +2550,37 @@ impl Processor {
                     }
                 }
                 StakeWithdrawSource::ValidatorRemoval => {
-                    if withdraw_lamports != stake_split_from.lamports() {
-                        msg!("Cannot withdraw a whole account worth {} lamports, must withdraw exactly {} lamports worth of pool tokens",
-                            withdraw_lamports, stake_split_from.lamports());
+                    let split_from_lamports = stake_split_from.lamports();
+                    // The upper bound for reasonable tolerance is twice the lamports per
+                    // pool token because we have two sources of rounding. The first happens
+                    // when reducing the stake account to as close to the minimum as possible,
+                    // and the second happens on this withdrawal.
+                    //
+                    // For example, if the pool token is extremely valuable, it might only
+                    // be possible to reduce the stake account to a minimum of
+                    // `stake_rent + minimum_delegation + lamports_per_pool_token - 1`.
+                    //
+                    // After that, the minimum amount of pool tokens to get to this amount
+                    // may actually be worth
+                    // `stake_rent + minimum_delegation + lamports_per_pool_token * 2 - 2`.
+                    // We give an extra grace on this check of two lamports, which should be
+                    // reasonable. At worst, it just means that a withdrawer is losing out
+                    // on two lamports.
```

Comment:
I am curious, I think there is something I am missing here. We are comparing against the actual lamports in the stake account here, so a tolerance of +lamports_per_token should always be enough. 

For example, let's assume we only reduce the stake account to minimum_stake_lamports + (lamports_per_token - 1). But then `split_from_lamports = minimum_stake_lamports +(lamports_per_token -  1)` already, and we can always find a token burn amount so that we are within +lamports_per_token of that. No need to double, as far as I understand.

Reply:
You're right, I just wrote the test poorly during my first pass. No need for the limit, thanks for taking the time to notice this!